pilot/pkg/networking/plugin/authn/authentication.go

```diff
@@ -92,13 +96,62 @@ func buildFilter(in *plugin.InputParams, mutable *networking.MutableObjects) err
 
 // OnInboundPassthrough is called whenever a new passthrough filter chain is added to the LDS output.
 func (Plugin) OnInboundPassthrough(in *plugin.InputParams, mutable *networking.MutableObjects) error {
-	return nil
+	if in.Node.Type != model.SidecarProxy {
+		// Only care about sidecar.
+		return nil
+	}
+
+	return buildFilter(in, mutable, true)
 }
 
 // OnInboundPassthroughFilterChains is called for plugin to update the pass through filter chain.
 func (Plugin) OnInboundPassthroughFilterChains(in *plugin.InputParams) []networking.FilterChain {
-	// Pass nil for ServiceInstance so that we never consider any alpha policy for the pass through filter chain.
 	applier := factory.NewPolicyApplier(in.Push, in.Node.Metadata.Namespace, labels.Collection{in.Node.Metadata.Labels})
-	// Pass 0 for endpointPort so that it never matches any port-level policy.
-	return applier.InboundFilterChain(0, constants.DefaultSdsUdsPath, in.Node, in.ListenerProtocol, trustDomainsForValidation(in.Push.Mesh))
+	trustDomains := trustDomainsForValidation(in.Push.Mesh)
+	// First generate the default passthrough filter chains, pass 0 for endpointPort so that it never matches any port-level policy.
+	filterChains := applier.InboundFilterChain(0, constants.DefaultSdsUdsPath, in.Node, in.ListenerProtocol, trustDomains)
+
+	// Then generate the per-port passthrough filter chains.
+	for port := range applier.PortLevelSetting() {
+		// Skip the per-port passthrough filterchain if the port is already handled by OnInboundFilterChains().
+		if portAlreadyHandled(port, in.Node) {
+			continue
+		}
+
+		log.Debugf("Generated extra passthrough filter chain for port: %d", port)
+		portLevelFilterChains := applier.InboundFilterChain(port, constants.DefaultSdsUdsPath, in.Node, in.ListenerProtocol, trustDomains)
+		for _, fc := range portLevelFilterChains {
+			// Set the port to distinguish from the default passthrough filter chain.
+			if fc.FilterChainMatch == nil {
+				fc.FilterChainMatch = &envoy_config_listener_v3.FilterChainMatch{}
+			}
+			fc.FilterChainMatch.DestinationPort = &wrappers.UInt32Value{Value: port}
+			filterChains = append(filterChains, fc)
+		}
+	}
+
+	return filterChains
+}
+
+func portAlreadyHandled(port uint32, node *model.Proxy) bool {
```

Comment:
Feel other naming can improve readability a bit. `requirePerPortPassthroughFilterChain` or something.

Currently, we need to read the implementation of this method to understand what does it mean 'already handled'.

Reply:
changed to needPerPortPassthroughFilterChain